enigma-principal/app/src/controller/km_controller.rs

```diff
@@ -0,0 +1,413 @@
+use std::path::PathBuf;
+use std::clone::Clone;
+
+use ethabi::{Log, RawLog};
+use failure::Error;
+use sgx_types::sgx_enclave_id_t;
+use web3::types::{H256, TransactionReceipt, U256, H160};
+use rustc_hex::ToHex;
+
+use enigma_tools_m::keeper_types::InputWorkerParams;
+use enigma_tools_u::web3_utils::enigma_contract::{ContractFuncs, ContractQueries, EnigmaContract};
+use enigma_tools_u::common_u::errors::Web3Error;
+use enigma_tools_u::esgx::equote::retry_quote;
+use enigma_tools_u::attestation_service::service::AttestationService;
+
+use controller::km_utils::KMConfig;
+use epochs::epoch_keeper_u::set_or_verify_worker_params;
+use esgx::equote;
+use common_u::errors::EpochStateTransitionErr;
+use common_u::custom_errors::{ControllerError, VerifierError, EnclaveError, EpochError};
+use epochs::verifier::EpochVerifier;
+use epochs::epoch_types::{EPOCH_STATE_UNCONFIRMED, SignedEpoch, WORKER_PARAMETERIZED_EVENT, WorkersParameterizedEvent};
+
+lazy_static!{ pub static ref GAS_LIMIT: U256 = 5_999_999.into(); }
+
+
+pub struct KMController {
+    pub contract: EnigmaContract,
+    pub epoch_verifier: EpochVerifier,
+    pub eid: sgx_enclave_id_t,
+    pub config: KMConfig,
+}
+
+impl KMController {
+    pub fn new(eid: sgx_enclave_id_t, dir_path: PathBuf, contract: EnigmaContract, config: KMConfig) -> Result<KMController, ControllerError> {
+        let epoch_verifier = EpochVerifier::new(dir_path).map_err(|e| ControllerError::VerifierError(e))?;
+        let controller = Self { contract, epoch_verifier, eid, config };
+        controller.verify_worker_params()?;
+        Ok(controller)
+    }
+
+    /// Find confirmed `EpochState` by block number
+    /// # Arguments
+    ///
+    /// * `block_number` - A block number in the desired epoch
+    pub fn find_epoch(&self, block_number: U256) -> Result<SignedEpoch, ControllerError> {
+        self.epoch_verifier.get_confirmed_by_block_number(block_number).map_err(|e| ControllerError::VerifierError(e))
+    }
+
+    /// Find the last confirmed `EpochState`
+    pub fn find_last_epoch(&self) -> Result<SignedEpoch, ControllerError> {
+        if self.epoch_verifier.is_last_unconfirmed().map_err(|e| ControllerError::VerifierError(e))? {
+            return Err(ControllerError::EpochError(EpochError::UnconfirmedState));
+        }
+        self.epoch_verifier.last(true).map_err(|e| ControllerError::VerifierError(e))
+    }
+
+    #[logfn(DEBUG)]
+    fn parse_worker_parameterized(&self, receipt: &TransactionReceipt) -> Result<Log, ControllerError> {
+        if receipt.logs.is_empty() {
+            // todo: change the web3error to a proper custom error
+            Err(Web3Error {
+                message: format!("A connection error occurred with the Smart Contract- workerParams did not return a log response" ),
+            })?
+        }
+        let log = receipt.logs[0].clone();
```

Comment:
This line always fails when a KM tries to connect to a network where another KM has already registered.
Please change it to something that will allow us to understand the problem when we see it in the future:
```suggestion
        let log = receipt.logs.get(0).ok_or(ControllerError::Something)?.clone();
```
Or catch the problem earlier in execution, if possible, in a way that will guarantee that this indexing never fails

Reply:
That's the reason I added the lines above